src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,153 @@
+            head = tail = newNode;
+        } else {
+            tail.next = newNode;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        checkIndex(index);
+        if (index == 0) {
+            Node<T> newNode = new Node<>(null, value, head);
+            newNode.next = head;
+            head = newNode;
+        } else {
+            Node<T> addedNode = getNodeByIndex(index);
+            Node<T> prevNode = addedNode.prev;
+            Node<T> newNode = new Node<>(prevNode, value, addedNode);
+            addedNode.prev = newNode;
+            if (prevNode == null) {
+                head = newNode;
+            } else {
+                prevNode.next = newNode;
+            }
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T item : list) {
+            add(item);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> newNode = getNodeByIndex(index);
+        T oldNode = newNode.item;
+        newNode.item = value;
+        return oldNode;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        return unlink(getNodeByIndex(index));
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> removedNode = getNodeByValue(object);
+        if (removedNode != null) {
+            unlink(removedNode);
+            return true;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return head == null;
+    }
+
+    private T unlink(Node<T> node) {
+        Node<T> first = node.prev;
+        Node<T> second = node.next;
+        if (first == null) {
+            head = second;
+        } else {
+            first.next = second;
+        }
+        if (second == null) {
+            tail = first;
+        } else {
+            second.prev = first;
+        }
+        T element = node.item;
+        node.item = null;
+        size--;
+        return element;
+    }
+
+    public Node<T> getNodeByIndex(int index) {
+        checkIndex(index);
+        Node<T> current = head;
+        if (index < (size / 2)) {
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+        } else {
+            current = tail;
+            for (int i = size - 1; i > index; i--) {
+                current = current.prev;
+            }
+        }
+        return current;
+    }
+
+    private Node<T> getNodeByValue(T value) {
+        for (int i = 0; i < size; i++) {
+            if ((getNodeByIndex(i).item == value)
```

Comment:
You use method getNodeByIndex() in **loop** and this method has **loop** inside too. You increase the complexity of the algorithm from O(n) to O(n^2). Let's create a new variable in the method getNodeByValue before the loop and will change it every iteration (_for example_ currentNode = currentNode.next)